libfwupdplugin/fu-progress.c

```diff
@@ -0,0 +1,774 @@
+/*
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuProgress"
+
+#include "config.h"
+
+#include "fu-progress.h"
+
+/**
+ * FuProgress:
+ *
+ * Objects can use fu_progress_set_percentage() if the absolute percentage
+ * is known. Percentages should always go up, not down.
+ *
+ * Modules usually set the number of steps that are expected using
+ * fu_progress_set_steps() and then after each section is completed,
+ * the fu_progress_step_done() function should be called. This will automatically
+ * call fu_progress_set_percentage() with the correct values.
+ *
+ * #FuProgress allows sub-modules to be "chained up" to the parent module
+ * so that as the sub-module progresses, so does the parent.
+ * The child can be reused for each section, and chains can be deep.
+ *
+ * To get a child object, you should use fu_progress_get_child() and then
+ * use the result in any sub-process. You should ensure that the child
+ * is not re-used without calling fu_progress_step_done().
+ *
+ * There are a few nice touches in this module, so that if a module only has
+ * one progress step, the child progress is used for updates.
+ *
+ *    static void
+ *    _do_something(FuProgress *self)
+ *    {
+ *       // setup correct number of steps
+ *       fu_progress_set_steps(self, 2);
+ *
+ *       // run a sub function
+ *       _do_something_else1(fu_progress_get_child(self));
+ *
+ *       // this section done
+ *       fu_progress_step_done(self);
+ *
+ *       // run another sub function
+ *       _do_something_else2(fu_progress_get_child(self));
+ *
+ *       // this progress done (all complete)
+ *       fu_progress_step_done(self);
+ *    }
+ *
+ * See also: [class@FuDevice]
+ */
+
+typedef struct {
+	gchar *id;
+	guint percentage;
+	FwupdStatus status;
+	GPtrArray *steps;
+	gboolean profile;
+	gdouble global_share;
+	GTimer *timer;
+	guint step_now;
+	guint step_max;
+	gulong percentage_child_id;
+	gulong status_child_id;
+	FuProgress *child;
+	FuProgress *parent; /* no-ref */
+} FuProgressPrivate;
+
+typedef struct {
+	FwupdStatus status;
+	guint value;
+	gdouble profile;
+} FuProgressStep;
+
+enum { SIGNAL_PERCENTAGE_CHANGED, SIGNAL_STATUS_CHANGED, SIGNAL_LAST };
+
+static guint signals[SIGNAL_LAST] = {0};
+
+G_DEFINE_TYPE_WITH_PRIVATE(FuProgress, fu_progress, G_TYPE_OBJECT)
+#define GET_PRIVATE(o) (fu_progress_get_instance_private(o))
+
+/**
+ * fu_progress_get_id:
+ * @self: a #FuProgress
+ *
+ * Return the id of the progress, which is normally set by the caller.
+ *
+ * Returns: progress ID
+ *
+ * Since: 1.7.0
+ **/
+const gchar *
+fu_progress_get_id(FuProgress *self)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_PROGRESS(self), NULL);
+	return priv->id;
+}
+
+/**
+ * fu_progress_set_id:
+ * @self: a #FuProgress
+ * @id: progress ID, normally `G_STRLOC`
```

Comment:
the main reason for this is to tell which plugin you're profiling right?  Considering this wouldn't it make more sense to just use `fu_plugin_get_name`?

Reply:
(And if so, then this can be more implicit during initialization)

Reply:
> the main reason for this is to tell which plugin you're profiling right

Actually it's more granular than that. So when you're profiling if you have a FuProgress with `75%, 25%` and then a child one of the first that has `10%, 90%` the profiler tells you actually _the parent one needs to be 30%, 70%_ and _the child one needs to be 50%, 50%_ -- i.e. you need more than the plugin name to know which one the profiler is referring to.